crates/rethnet_eth/src/remote/cacheable_method_invocation.rs

```diff
@@ -181,22 +253,254 @@ impl<'a> TryFrom<&'a MethodInvocation> for CacheableMethodInvocation<'a> {
             | MethodInvocation::EvmSetAutomine(_)
             | MethodInvocation::EvmSetIntervalMining(_)
             | MethodInvocation::EvmSetNextBlockTimestamp(_)
-            | MethodInvocation::EvmSnapshot() => Err(MethodNotCacheableError(value.clone())),
+            | MethodInvocation::EvmSnapshot() => {
+                Err(MethodNotCacheableError::MethodInvocation(value.clone()))
+            }
+        }
+    }
+}
+
+/// A block argument specification that is cacheable.
+#[derive(Clone, Debug)]
+pub(super) enum CacheableBlockSpec<'a> {
+    /// Block number
+    Number { block_number: &'a U256 },
+    /// Block hash
+    Hash {
+        block_hash: &'a B256,
+        require_canonical: Option<bool>,
+    },
+    /// "earliest" block tag
+    Earliest,
+    /// "safe" block tag
+    Safe,
+    /// "finalized" block tag
+    Finalized,
```

Comment:
> We shouldn't be caching a key with the tags themselves, but instead resolving to the appropriate Number, as the tags are "moving targets"

Yes, it’s handled downstream from here.

We can turn a `MethodInvocation` into `CacheableMethodInvocation` if the latter is potentially cacheable. This is why we need potentially cacheable symbolic block specs (“safe”, “finalized” and “earliest”) to be part of `CacheableBlockSpec`. 

Then we distinguish between `ReadCacheKey` and `WriteCacheKey`. 

A `WriteCacheKey` is an enum that can be either 
- fully resolved, 
- need a check that its block number is safe to cache, 
- or if it contains a symbolic block spec, it needs to be resolved to a block number from the response.

`ReadCacheKey` will be `None` if the block tag is symbolic, but it’ll produce a cache key for potentially unsafe block numbers. The reason for this is that it’ll be a cache miss since we don’t write to it, but we save a request to fetch the latest block number before reading.

---

I made the docstring a bit clearer for the `CacheableBlockSpec` enum: [83d8b4f](https://github.com/NomicFoundation/hardhat/pull/4319/commits/83d8b4f410e7fc2a68e8fa535a0ac0e330f536fa)

Reply:
Does it make sense to have a difference between the actual cache type and a request to be cached type?

I'm asking because I like having statically encoded types that avoid any room for error, but you know the code better than me to be able to judge that.

Reply:
Yes, the way it's written now, it's impossible to construct a cache key without checking the invariants. 👍 

We need some sort of data structure to hold data related to requests that are potentially cacheable for two reasons:

1. The hash calculation operates on this data structure.
2. There is a back and forth where this data is passed as part of an abstraction to resolve some data.

We could make this a bit cleaner by introducing two separate structures for the two cases, but it would be an overkill imo, as the types are local to the module.

Reply:
I made `CacheableMethodInvocation` and `CacheableBlockSpec` private: [f7fc5ce](https://github.com/NomicFoundation/hardhat/pull/4319/commits/f7fc5ceb980955fca2307ccc5b2e6d68614e33e8). It makes things cleaner and removes the potential for misuse of the `CacheableBlockSpec`